### PR TITLE
HTTP 302 Found - redirrect support for getWOPIFileInfo

### PR DESCRIPTION
**Pull request by eszkadev**
With some loadbalancers it may happen that HTTP 302 Found
response with redirect location will appear.

### CODE REVIEW

**Review by Ashod — APPROVED** [review]
Looks good, thanks Szymon.

File: test/UnitWOPIHttpRedirectLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        LOG_INF("Fake wopi host request URI [" << uriReq.toString() << "]:\n");
+
+        // CheckFileInfo - always returns redirect response
+        if (request.getMethod() == "GET" && regInfo.match(uriReq.getPath()))
+        {
+            LOG_INF("Fake wopi host request, handling CheckFileInfo");
+
+            assertCheckFileInfoRequest(request);
+
+            LOK_ASSERT_MESSAGE("It is expected to stop requesting after 21 redirections", fileId <= 22);
```
Comment (Ashod) [inline test/UnitWOPIHttpRedirectLoop.cpp:55]: Yet another reason why magic numbers are bad: this test will either break, or become useless, when the redirctLimit default changes in the code. Why can't we have this as a compile-time const with a proper name and use that both in the code and tests?

**Review by Ashod — COMMENTED** [review]
Thanks Szymon, this is certainly an improvement that should help. I have some concerns about certain details, though. They might not block a release, but certainly worth improving things sooner rather than later.

**Comment by eszkadev** [comment]
I will introduce updates as follow up change

File: test/UnitWOPIHttpRedirectLoop.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+class UnitWopiHttpRedirectLoop : public WopiTestServer
+{
+    enum class Phase
+    {
+        Load,
+        Redirected,
+        Polling
+    } _phase;
+
+    const std::string params = "access_token=anything";
```
[the remark this answers is not in the document]
Comment (eszkadev) [reply test/UnitWOPIHttpRedirectLoop.cpp:27]: not possible: https://stackoverflow.com/questions/27123306/is-it-possible-to-use-stdstring-in-a-constexpr

File: test/UnitWOPIHttpRedirectLoop.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+            LOK_ASSERT_MESSAGE("Expected to be in Phase::Load or Phase::Redirected", _phase == Phase::Load || _phase == Phase::Redirected);
+            _phase = Phase::Redirected;
+
+            if (fileId == 22)
+                _phase = Phase::Polling;
+
+            std::ostringstream oss;
+            oss << "HTTP/1.1 302 Found\r\n"
+                "Location: " << helpers::getTestServerURI() << redirectUri << fileId++ << "?" << params << "\r\n"
```
[the remark this answers is not in the document]
Comment (eszkadev) [reply test/UnitWOPIHttpRedirectLoop.cpp:65]: done

File: wsd/Storage.cpp [hunk 4]
```diff
@@ -670,6 +671,26 @@ std::unique_ptr<WopiStorage::WOPIFileInfo> WopiStorage::getWOPIFileInfo(const Au
...
+                const std::string& location = httpResponse->get("Location");
+                LOG_TRC("WOPI::CheckFileInfo redirect to URI [" << LOOLWSD::anonymizeUrl(location) << "]:\n");
+
+                Poco::URI redirectUriObject(location);
+                setUri(redirectUriObject);
+                return getWOPIFileInfoForUri(redirectUriObject, auth, cookies, lockCtx, redirectLimit - 1);
+            }
+            else
+            {
+                LOG_TRC("WOPI::CheckFileInfo redirected too many times\n");
```
[the remark this answers is not in the document]
Comment (eszkadev) [reply wsd/Storage.cpp:690]: done

File: wsd/Storage.hpp [hunk 5]
```diff
@@ -625,7 +632,8 @@ class WopiStorage : public StorageBase
     /// Download the document from the given URI.
     /// Does not add authorization tokens or any other logic.
     std::string downloadDocument(const Poco::URI& uriObject, const std::string& uriAnonym,
-                                 const Authorization& auth, const std::string& cookies);
+                                 const Authorization& auth, const std::string& cookies,
+                                 unsigned redirectLimit = 21);
```
[the remark this answers is not in the document]
Comment (eszkadev) [reply wsd/Storage.hpp:636]: done

File: test/UnitWOPIHttpRedirectLoop.cpp (hunk 1 above)
Comment (eszkadev) [reply test/UnitWOPIHttpRedirectLoop.cpp:55]: done

**Comment by eszkadev** [comment]
updates in: https://github.com/CollaboraOnline/online/pull/2710

**Pull request merged**